# NLP Basics: (Natural Language processing)
- Applications of NLP:
- 1. Social Media posts
  2. Review/Comments Analysis
  3. Literature Analysis

In [3]:
text = "Happy to share the news that India Successfully landed on Moon.CHANDRAYANA - 3 was a complete success"
text

'Happy to share the news that India Successfully landed on Moon.CHANDRAYANA - 3 was a complete success'

In [4]:
# 1. Convert the entire text into a lower case. small letters.
# India and india are two different things. Notice that "i" in India is capital and "i" in india is small.

text1 = text.lower()
text1

'happy to share the news that india successfully landed on moon.chandrayana - 3 was a complete success'

In [8]:
t = "It is a Truth Universally acknowledged, that a single man in possession of a #good fortune, must be in want of a @Good wife!"
t

'It is a Truth Universally acknowledged, that a single man in possession of a #good fortune, must be in want of a @Good wife!'

In [9]:
t1 = t.lower()
t1

'it is a truth universally acknowledged, that a single man in possession of a #good fortune, must be in want of a @good wife!'

# Notes:
- The idea of converting a text to a lower case is to make python understand that good and Good both give the same meaning.

# Removing Punctuations:


In [10]:
import string
print(string.punctuation)

!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~


# How to remove punctuations from the text
1. Create an empty list - To catch all the useful words from the text.(While looping through the text word by word, if a punctuation    
is encountered ignore it).
2. use a for loop to visit each char from the text.
3. Put a condition on char that if a char is present in string.punctuation ignore it.
4. use the join method on the empty list created above to get a text without punctuations

In [12]:
t

'It is a Truth Universally acknowledged, that a single man in possession of a #good fortune, must be in want of a @Good wife!'

In [15]:
for char in t:
    print(char)

I
t
 
i
s
 
a
 
T
r
u
t
h
 
U
n
i
v
e
r
s
a
l
l
y
 
a
c
k
n
o
w
l
e
d
g
e
d
,
 
t
h
a
t
 
a
 
s
i
n
g
l
e
 
m
a
n
 
i
n
 
p
o
s
s
e
s
s
i
o
n
 
o
f
 
a
 
#
g
o
o
d
 
f
o
r
t
u
n
e
,
 
m
u
s
t
 
b
e
 
i
n
 
w
a
n
t
 
o
f
 
a
 
@
G
o
o
d
 
w
i
f
e
!


In [16]:
emp_lst = []
for char in t:
    if char not in string.punctuation:
        emp_lst.append(char)

print(emp_lst)

['I', 't', ' ', 'i', 's', ' ', 'a', ' ', 'T', 'r', 'u', 't', 'h', ' ', 'U', 'n', 'i', 'v', 'e', 'r', 's', 'a', 'l', 'l', 'y', ' ', 'a', 'c', 'k', 'n', 'o', 'w', 'l', 'e', 'd', 'g', 'e', 'd', ' ', 't', 'h', 'a', 't', ' ', 'a', ' ', 's', 'i', 'n', 'g', 'l', 'e', ' ', 'm', 'a', 'n', ' ', 'i', 'n', ' ', 'p', 'o', 's', 's', 'e', 's', 's', 'i', 'o', 'n', ' ', 'o', 'f', ' ', 'a', ' ', 'g', 'o', 'o', 'd', ' ', 'f', 'o', 'r', 't', 'u', 'n', 'e', ' ', 'm', 'u', 's', 't', ' ', 'b', 'e', ' ', 'i', 'n', ' ', 'w', 'a', 'n', 't', ' ', 'o', 'f', ' ', 'a', ' ', 'G', 'o', 'o', 'd', ' ', 'w', 'i', 'f', 'e']


In [17]:
emp_lst = []
for char in t:
    if char not in string.punctuation:
        emp_lst.append(char)

t1 = "".join(emp_lst)
t1

'It is a Truth Universally acknowledged that a single man in possession of a good fortune must be in want of a Good wife'

In [44]:
t2_punc_rem = "".join([char for char in t if char not in string.punctuation])
t2_punc_rem

'It is a Truth Universally acknowledged that a single man in possession of a good fortune must be in want of a Good wife'

In [21]:
t1

'It is a Truth Universally acknowledged that a single man in possession of a good fortune must be in want of a Good wife'

In [22]:
len(t1)

119

# Notes:
- 119 is the count of all the individual characters along with the spaces.
-  I is 1, t is 2, " " is 3, i is 4, s is 5, " " is 6 and so on

In [24]:
t2 = t1.split(" ")
t2

['It',
 'is',
 'a',
 'Truth',
 'Universally',
 'acknowledged',
 'that',
 'a',
 'single',
 'man',
 'in',
 'possession',
 'of',
 'a',
 'good',
 'fortune',
 'must',
 'be',
 'in',
 'want',
 'of',
 'a',
 'Good',
 'wife']

In [25]:
len(t2)

24

# Notes:
- 24 is the count of all words present in the text. Here spaces are not considered in the total count

In [27]:
t3 = set(t2)
t3

{'Good',
 'It',
 'Truth',
 'Universally',
 'a',
 'acknowledged',
 'be',
 'fortune',
 'good',
 'in',
 'is',
 'man',
 'must',
 'of',
 'possession',
 'single',
 'that',
 'want',
 'wife'}

In [28]:
len(t3)

19

In [30]:
t4 = set(t1.lower().split(" "))
t4

{'a',
 'acknowledged',
 'be',
 'fortune',
 'good',
 'in',
 'is',
 'it',
 'man',
 'must',
 'of',
 'possession',
 'single',
 'that',
 'truth',
 'universally',
 'want',
 'wife'}

In [31]:
len(t4)

18

# Notes:
- To get the unique words from the text use set() and len().

# Installing nltk library

In [18]:
!pip install nltk

Error processing line 3 of C:\Users\AMANT\anaconda3\lib\site-packages\googleapis_common_protos-1.56.4-py3.10-nspkg.pth:

  Traceback (most recent call last):
    File "C:\Users\AMANT\anaconda3\lib\site.py", line 169, in addpackage
      exec(line)
    File "<string>", line 1, in <module>
    File "<frozen importlib._bootstrap>", line 562, in module_from_spec
  AttributeError: 'NoneType' object has no attribute 'loader'

Remainder of file ignored


In [19]:
import nltk

In [32]:
t1

'It is a Truth Universally acknowledged that a single man in possession of a good fortune must be in want of a Good wife'

In [33]:
from nltk.tokenize import word_tokenize
text_tokenized = word_tokenize(t1)
text_tokenized

['It',
 'is',
 'a',
 'Truth',
 'Universally',
 'acknowledged',
 'that',
 'a',
 'single',
 'man',
 'in',
 'possession',
 'of',
 'a',
 'good',
 'fortune',
 'must',
 'be',
 'in',
 'want',
 'of',
 'a',
 'Good',
 'wife']

# Notes:
- Tokenizing in NLP means extracting a words(series of characters) from the text.
- The output will be a list of words(tokens) in a list

In [34]:
text5 = "This is a NLP basics session. This is a NLP basics session. NLP is a nice area of research. I am learning to analyze text data"
text5

'This is a NLP basics session. This is a NLP basics session. NLP is a nice area of research. I am learning to analyze text data'

In [35]:
from nltk import sent_tokenize
sent_tokenize(text5)

['This is a NLP basics session.',
 'This is a NLP basics session.',
 'NLP is a nice area of research.',
 'I am learning to analyze text data']

# Notes:
- The output of a sent_tokenize will be a list of sentences(collection of tokens).
  

# Stops Words filtering:
- Stop Words are connecting words used in a sentence which individually do not have any meaning. eg: (the, and, this, were, etc.)
  

In [38]:
from nltk.corpus import stopwords
stop_words = stopwords.words("english")
print(stop_words)

['i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', "you're", "you've", "you'll", "you'd", 'your', 'yours', 'yourself', 'yourselves', 'he', 'him', 'his', 'himself', 'she', "she's", 'her', 'hers', 'herself', 'it', "it's", 'its', 'itself', 'they', 'them', 'their', 'theirs', 'themselves', 'what', 'which', 'who', 'whom', 'this', 'that', "that'll", 'these', 'those', 'am', 'is', 'are', 'was', 'were', 'be', 'been', 'being', 'have', 'has', 'had', 'having', 'do', 'does', 'did', 'doing', 'a', 'an', 'the', 'and', 'but', 'if', 'or', 'because', 'as', 'until', 'while', 'of', 'at', 'by', 'for', 'with', 'about', 'against', 'between', 'into', 'through', 'during', 'before', 'after', 'above', 'below', 'to', 'from', 'up', 'down', 'in', 'out', 'on', 'off', 'over', 'under', 'again', 'further', 'then', 'once', 'here', 'there', 'when', 'where', 'why', 'how', 'all', 'any', 'both', 'each', 'few', 'more', 'most', 'other', 'some', 'such', 'no', 'nor', 'not', 'only', 'own', 'same', 'so', 'than', '

In [39]:
len(stop_words)

179

## Stopword Removal
1. Define an empty list - Collect all the words expect stop words
2. Perform word tokenization - To get a list of all words from the text.
3. Use for loop to extract each token from the list of text from step 2.
4. Use a condition to ignore the stop words appearing in stopwords.words("english")
5. Use the join method on the empty list to get a text without stop words

In [42]:
lst = []
for word in word_tokenize(t1):
    if word not in stopwords.words("english"):
        lst.append(word)

t1_modified = " ".join(lst)
print(t1_modified)

It Truth Universally acknowledged single man possession good fortune must want Good wife


In [43]:
t2_modeified = " ".join([word for word in word_tokenize(t1) if word not in stopwords.words("english")])
t2_modeified

'It Truth Universally acknowledged single man possession good fortune must want Good wife'

In [ ]:
lst = []
for
    if
        lst.append(word)

In [ ]:
[word for if]

In [45]:
t1

'It is a Truth Universally acknowledged that a single man in possession of a good fortune must be in want of a Good wife'

In [46]:
nltk.FreqDist(t1)

FreqDist({' ': 23, 'o': 10, 'a': 9, 'n': 9, 's': 8, 'e': 8, 't': 7, 'i': 7, 'l': 4, 'd': 4, ...})

# Notes:
- The total count of each char appearing in the text is displayed
- The output is already present in sorted order
- When the count is same it will display the char in alphabetical order

In [47]:
nltk.FreqDist(t1.split(" "))

FreqDist({'a': 4, 'in': 2, 'of': 2, 'It': 1, 'is': 1, 'Truth': 1, 'Universally': 1, 'acknowledged': 1, 'that': 1, 'single': 1, ...})

# Notes:
- Total count of each token appearing in the text
- Output is in sorted order and if the count is same the it should be in alphabetical order 